README.md
# Asgard app stats collector  [![Build Status](https://travis-ci.org/B2W-BIT/asgard-app-stats-collector.svg?branch=master)](https://travis-ci.org/B2W-BIT/asgard-app-stats-collector) [![codecov](https://codecov.io/gh/B2W-BIT/asgard-app-stats-collector/branch/master/graph/badge.svg)](https://codecov.io/gh/B2W-BIT/asgard-app-stats-collector)


Coletor de uso de CPU e RAM para todas tasks do cluster do Asgard.


## Fluxo de coleta

* Conecta no master
* Descobre o IP de **todos** os slaves do cluster
* Conecta em cada slave, no endpoint `/monitor/statistics.json`
* Posta o JSON resultante no RabbitMQ


## Configurações


```
STATS_COLLECTOR_RABBITMQ_HOST: IP do RabbitMQ onde as mensagens serão colocadas;
STATS_COLLECTOR_RABBITMQ_PWD: Senha usada no RabbitMQ
STATS_COLLECTOR_RABBITMQ_USER: Username usado no RabbitMQ
STATS_COLLECTOR_RABBITMQ_VHOST: Virtual Host do RabbitMQ
STATS_COLLECTOR_RABBITMQ_RK: Routing Key usada para depositar as mensagens. O exchange usado é sempre `""`
STATS_COLLECTOR_MESOS_MASTER_IP: IP do meso Master (sem incluir a porta)
STATS_COLLECTOR_REDIS_URL: Url completa para o redis
STATS_COLLECTOR_REDIS_POOL_MIN: Tamanho inicial do pool de coexões com o redis
STATS_COLLECTOR_REDIS_POOL_MAX: Tamanho máximo do pool de conexões com o redis
STATS_COLLECTOR_INCLUDE_RAW_METRICS: [0|1] Default: 0. Indica de os valores brutos, usados nos cálculos também serão incluídos no documento final.
STATS_CONNECT_TIMEOUT: Default: 2s Timeout em segundos para as conexões HTTP
STATS_REQUEST_TIMEOUT: Default: 5s Timeout em segundos para os requests HTTP (completos, incluindo tempo para receber a response)
```

## Trabalhos futuros

Migra o código para que consiga se conectar em um cluster de mesos master, ou seja, se alguns dos
masters não responder, vai nos próximos.

Isso significa que vamos usar o `get_option()` que existe na `asgard-api-sdk`.

HOLLOWMAN_MESOS_ADDRESS_N: Lista com IP:PORTA de todos os mesos masters

Exemplo:
```
HOLLOWMAN_MESOS_ADDRESS_0: http://IP1:PORTA1
HOLLOWMAN_MESOS_ADDRESS_1: http://IP2:PORTA2
HOLLOWMAN_MESOS_ADDRESS_2: http://IP3:PORTA3
```

* Escrever mais testes, aumentar a cobertura de testes.
* Tratar melhor casos de errot (ConnectionError, ConnectionTimeout, etc)



CHANGELOG.md
# Changelog


## 0.5.0
 * Atualizado aiologger pra 0.5.0

## 0.3.0
 * Atualizando dependencias;

## 0.2.1
 * Fix na extração do nome da app a partir do nome da task;

## 0.2.0

 * Mudança no nome de todas as envs, trocando o prefixo `STATUS` por `STATS`;
 * Adição do cálculo correto de uso de CPU e CPU throttling;


tests/test_conf.py
import os
from importlib import reload

import asynctest
from asynctest import mock

import status_collector
from status_collector import conf, main


class ConfTest(asynctest.TestCase):
    async def test_read_redis_configs(self):
        expected_redis_url = "redis://cache.asgard/0"
        with mock.patch.dict(
            os.environ,
            STATS_COLLECTOR_REDIS_URL=expected_redis_url,
            STATS_COLLECTOR_REDIS_POOL_MIN="10",
            STATS_COLLECTOR_REDIS_POOL_MAX="32",
        ), mock.patch.object(
            conf.aioredis, "create_redis_pool"
        ) as create_redis_pool_mock:
            reload(conf)
            self.assertEqual(expected_redis_url, conf.REDIS_URL)
            self.assertEqual(10, conf.REDIS_POOL_MIN)
            self.assertEqual(32, conf.REDIS_POOL_MAX)

    def test_read_rabbitmq_configs(self):

        with mock.patch.dict(
            os.environ,
            STATS_COLLECTOR_RABBITMQ_HOST="rmq.asgard",
            STATS_COLLECTOR_RABBITMQ_USER="guest",
            STATS_COLLECTOR_RABBITMQ_PWD="guest-pwd",
            STATS_COLLECTOR_RABBITMQ_RK="routing_key",
            STATS_COLLECTOR_RABBITMQ_VHOST="vhost",
        ):
            reload(conf)
            self.assertEqual("rmq.asgard", conf.STATS_COLLECTOR_RABBITMQ_HOST)
            self.assertEqual("guest", conf.STATS_COLLECTOR_RABBITMQ_USER)
            self.assertEqual("guest-pwd", conf.STATS_COLLECTOR_RABBITMQ_PWD)
            self.assertEqual("routing_key", conf.STATS_COLLECTOR_RABBITMQ_RK)
            self.assertEqual("vhost", conf.STATS_COLLECTOR_RABBITMQ_VHOST)

    def test_read_mesos_master_address(self):
        expected_mesos_master_ip = "192.168.44.89"
        with mock.patch.dict(
            os.environ, STATS_COLLECTOR_MESOS_MASTER_IP=expected_mesos_master_ip
        ):
            reload(conf)
            self.assertEqual(
                expected_mesos_master_ip, conf.STATS_COLLECTOR_MESOS_MASTER_IP
            )

    def test_read_raw_metrics_configs(self):
        reload(conf)
        self.assertFalse(conf.STATS_COLLECTOR_INCLUDE_RAW_METRICS)

        with mock.patch.dict(
            os.environ, STATS_COLLECTOR_INCLUDE_RAW_METRICS="1"
        ):
            reload(conf)
            self.assertTrue(conf.STATS_COLLECTOR_INCLUDE_RAW_METRICS)

        with mock.patch.dict(
            os.environ, STATS_COLLECTOR_INCLUDE_RAW_METRICS="0"
        ):
            reload(conf)
            self.assertFalse(conf.STATS_COLLECTOR_INCLUDE_RAW_METRICS)

    def test_read_connection_timeout_settings(self):

        with mock.patch.dict(
            os.environ, STATS_CONNECT_TIMEOUT="5", STATS_REQUEST_TIMEOUT="30",
        ):
            reload(conf)
            self.assertEqual(5, conf.STATS_CONNECT_TIMEOUT)
            self.assertEqual(30, conf.STATS_REQUEST_TIMEOUT)


tests/test_main.py
from importlib import reload
import asynctest
from asynctest import mock
from status_collector import conf, main
import status_collector
import os

class MainTest(asynctest.TestCase):

    async def test_main_creates_redis_pool(self):
        expected_redis_url = "redis://cache.asgard/0"
        with mock.patch.dict(os.environ,
                             STATS_COLLECTOR_REDIS_URL=expected_redis_url,
                             STATS_COLLECTOR_REDIS_POOL_MIN="10",
                             STATS_COLLECTOR_REDIS_POOL_MAX="32"), \
                mock.patch.object(status_collector.aioredis, "create_redis_pool") as create_redis_pool_mock, \
                mock.patch.object(status_collector, "fetch_app_stats"):
            reload(conf)
            await main()
            create_redis_pool_mock.assert_awaited_with(expected_redis_url, minsize=10, maxsize=32)
            self.assertTrue(conf.cache)


tests/test_fetch_master.py
import asyncio

import asynctest
from aiologger.loggers.json import JsonLogger
from aioresponses import aioresponses
from asynctest import mock
from asynctest.mock import CoroutineMock

import status_collector
from status_collector import (
    get_slave_ip_list,
    get_slave_statistics,
    send_slave_statistics_to_queue,
    async_tasks,
    fetch_app_stats,
    build_statistic_for_response,
    conf,
)

slaves = {
    "slaves": [
        {
            "id": "90ba2028-fed0-4f19-a2b1-cb35baf9b2cd-S0",
            "hostname": "10.0.111.32",
            "port": 5051,
            "attributes": {
                "dc": "bit",
                "owner": "sieve",
                "mesos": "slave",
                "workload": "general",
            },
            "pid": "slave(1)@10.168.200.94:5051",
            "registered_time": 1528216448.5751,
        },
        {
            "id": "90ba2028-fed0-4f19-a2b1-cb35baf9b2cd-S0",
            "hostname": "10.0.111.33",
            "port": 5051,
            "attributes": {
                "dc": "bit",
                "owner": "sieve",
                "mesos": "slave",
                "workload": "general",
            },
            "pid": "slave(1)@10.168.200.94:5051",
            "registered_time": 1528216448.5751,
        },
        {
            "id": "90ba2028-fed0-4f19-a2b1-cb35baf9b2cd-S0",
            "hostname": "10.0.111.34",
            "port": 5051,
            "attributes": {
                "dc": "bit",
                "owner": "sieve",
                "mesos": "slave",
                "workload": "general",
            },
            "pid": "slave(1)@10.168.200.94:5051",
            "registered_time": 1528216448.5751,
        },
    ]
}

slave_statistics_response_mock = [
    {
        "executor_id": "sieve_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430",
        "executor_name": "Command Executor (Task: sieve_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430) (Command: NO EXECUTABLE)",
        "framework_id": "27b52920-3899-4b90-a1d6-bf83a87f3612-0000",
        "source": "sieve_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430",
        "statistics": {
            "cpus_limit": 8.25,
            "cpus_nr_periods": 769021,
            "cpus_nr_throttled": 1046,
            "cpus_system_time_secs": 34501.45,
            "cpus_throttled_time_secs": 352.597023453,
            "cpus_user_time_secs": 96348.84,
            "mem_anon_bytes": 4845449216,
            "mem_file_bytes": 260165632,
            "mem_limit_bytes": 7650410496,
            "mem_mapped_file_bytes": 7159808,
            "mem_rss_bytes": 5105614848,
            "timestamp": 1388534400.0,
        },
    }
]

slave_statistics_response_mock_multiple_tasks = [
    {
        "executor_id": "sieve_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430",
        "executor_name": "Command Executor (Task: sieve_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430) (Command: NO EXECUTABLE)",
        "framework_id": "27b52920-3899-4b90-a1d6-bf83a87f3612-0000",
        "source": "sieve_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430",
        "statistics": {
            "cpus_limit": 8.25,
            "cpus_nr_periods": 769021,
            "cpus_nr_throttled": 1046,
            "cpus_system_time_secs": 34501.45,
            "cpus_throttled_time_secs": 352.597023453,
            "cpus_user_time_secs": 96348.84,
            "mem_anon_bytes": 4845449216,
            "mem_file_bytes": 260165632,
            "mem_limit_bytes": 7650410496,
            "mem_mapped_file_bytes": 7159808,
            "mem_rss_bytes": 5105614848,
            "timestamp": 1388534400.0,
        },
    },
    {
        "executor_id": "sieve2_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430",
        "executor_name": "Command Executor (Task: sieve_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430) (Command: NO EXECUTABLE)",
        "framework_id": "27b52920-3899-4b90-a1d6-bf83a87f3612-0000",
        "source": "sieve_captura_kirby_powerup.947f6124-762a-11e8-b78b-c678b0156430",
        "statistics": {
            "cpus_limit": 8.25,
            "cpus_nr_periods": 769021,
            "cpus_nr_throttled": 1046,
            "cpus_system_time_secs": 34501.45,
            "cpus_throttled_time_secs": 352.597023453,
            "cpus_user_time_secs": 96348.84,
            "mem_anon_bytes": 4845449216,
            "mem_file_bytes": 260165632,
            "mem_limit_bytes": 7650410496,
            "mem_mapped_file_bytes": 7159808,
            "mem_rss_bytes": 5105614848,
            "timestamp": 1388534400.0,
        },
    },
]


class FecthMasterTest(asynctest.TestCase):
    def setUp(self):
        self.loggerMock = self.loggerMock = asynctest.Mock(
            spec=JsonLogger, exception=CoroutineMock()
        )

    async def test_get_slaves_ips_list(self):
        with aioresponses() as m:
            m.get("http://10.11.43.96:5050/slaves", payload=slaves)
            slave_ips = await get_slave_ip_list("10.11.43.96", self.loggerMock)
        self.assertEqual(
            slave_ips,
            ["10.0.111.32:5051", "10.0.111.33:5051", "10.0.111.34:5051"],
        )

    async def test_slave_dont_exists(self):
        with aioresponses() as m:
            with self.assertRaises(Exception):
                m.get(
                    "http://10.0.111.32:5051/monitor/statistics.json",
                    exception=Exception("Invalid slave ip 10.0.111.32:5051."),
                )
                slave_statistics = await get_slave_statistics(
                    "10.0.111.32:5051", self.loggerMock
                )

    async def test_reraises_exception_if_error_on_fetch_slaves_ip_list(self):
        with aioresponses() as m:
            m.get(
                "http://10.0.1.1:5050/slaves",
                exception=Exception(
                    "Error fetching Slaves IP list 10.0.1.1:5051."
                ),
            )
            with self.assertRaises(Exception):
                slave_statistics = await get_slave_ip_list(
                    "10.0.1.1", self.loggerMock
                )

            self.loggerMock.exception.assert_awaited_with(
                {
                    "action": "get_slave_ip_list",
                    "exception": {
                        "message": "Error fetching Slaves IP list 10.0.1.1:5051.",
                        "traceback": mock.ANY,
                        "type": "Exception",
                    },
                    "master_ip": "10.0.1.1",
                }
            )

    async def test_putting_slave_statistics_on_rabbitMQ(self):
        slave_statistics = [
            {"taskname": "/infra/task"},
            {"taskname": "/infra/othertask"},
        ]
        queue = asynctest.mock.CoroutineMock(
            put=asynctest.mock.CoroutineMock(),
            connect=asynctest.mock.CoroutineMock(),
        )
        await send_slave_statistics_to_queue(
            slave_statistics, queue, self.loggerMock
        )
        self.assertEqual(
            [
                asynctest.mock.call(
                    body=slave_statistics[0],
                    routing_key=conf.STATS_COLLECTOR_RABBITMQ_RK,
                ),
                asynctest.mock.call(
                    body=slave_statistics[1],
                    routing_key=conf.STATS_COLLECTOR_RABBITMQ_RK,
                ),
            ],
            queue.put.await_args_list,
        )


calc-cpu.py
import asyncio
import aiohttp

import json
import os

agent_ip = os.environ.get("AGENT_IP")


CACHE = {}

def print_json(before, now):
    now_stats = now['statistics']
    before_stats = before['statistics']
    period_secs = now_stats['timestamp'] - before_stats['timestamp']
    cpu_limit = now_stats['cpus_limit']

    cpu_thr_secs = now_stats['cpus_throttled_time_secs'] - before_stats['cpus_throttled_time_secs']
    cpu_usr_secs = now_stats['cpus_user_time_secs'] - before_stats['cpus_user_time_secs']
    cpu_sys_secs = now_stats['cpus_system_time_secs'] - before_stats['cpus_system_time_secs']

    cpu_usr_host_pct = cpu_usr_secs / period_secs * 100
    cpu_sys_host_pct = cpu_sys_secs / period_secs * 100

    mem_bytes = now_stats['mem_rss_bytes']
    mem_pct = now_stats['mem_rss_bytes'] / now_stats['mem_limit_bytes'] * 100

    cpu_usr_pct = cpu_usr_host_pct / cpu_limit
    cpu_sys_pct = cpu_sys_host_pct / cpu_limit
    cpu_thr_pct = cpu_thr_secs / period_secs * 100
    cpu_pct = (cpu_usr_secs + cpu_sys_secs) / cpu_limit * 100

    data = {
        "stats": {
            "before": before['statistics'],
            "now": now['statistics']
        },
        "cpu_limit": cpu_limit,
        "host": agent_ip,
        "taskname": now['source'],
        "appname": "/infra/stress",
        "period_secs": period_secs,
        "cpu_usr_secs": cpu_usr_secs,
        "cpu_sys_secs": cpu_sys_secs,
        "cpu_thr_secs": cpu_thr_secs,
        "mem_bytes": mem_bytes,
        "mem_pct": mem_pct,
        "cpu_sys_host_pct": cpu_sys_host_pct,
        "cpu_usr_host_pct": cpu_usr_host_pct,
        "cpu_usr_pct": cpu_usr_pct,
        "cpu_sys_pct": cpu_sys_pct,
        "cpu_pct": cpu_pct,
        "cpu_thr_pct": cpu_thr_pct
    }

    print(json.dumps(data, indent=2))


async def main():
    while True:
        s = aiohttp.ClientSession()
        r = await s.get(f"http://{agent_ip}:5051/monitor/statistics.json")
        data = await r.json()
        await s.close()
        for task_data in data:
            task_name = task_data['source']
            if task_name not in CACHE:
                CACHE[task_name] = task_data
                continue
            before = CACHE[task_name]
            now = task_data
            now['statistics']['cpus_limit'] -= 0.1
            delta_t = now['statistics']['timestamp'] - before['statistics']['timestamp']
            cpu_usr_secs = now['statistics']['cpus_user_time_secs'] - before['statistics']['cpus_user_time_secs']
            cpu_sys_secs = now['statistics']['cpus_system_time_secs'] - before['statistics']['cpus_system_time_secs']
            cpu_thr_secs = now['statistics']['cpus_throttled_time_secs'] - before['statistics']['cpus_throttled_time_secs']


            cpu_usr_pct = cpu_usr_secs / delta_t * 100
            cpu_thr_pct = cpu_thr_secs / delta_t * 100
            cpu_usr_pct_mesos = cpu_usr_pct / now['statistics']['cpus_limit']

#            print (f"{task_name} elapsed={delta_t:.3f}, \
#usr_secs={cpu_usr_secs:.3f}, usr_pct={cpu_usr_pct:.3f}%, usr_pct_meso={cpu_usr_pct_mesos:.3f}% ({cpu_usr_pct/100:.3f}, {now['statistics']['cpus_limit']:.3f}), \
#thr_secs={cpu_thr_secs:.3f}, thr_pct={cpu_thr_pct:.3f}%,\
#")

            print_json(before, now)
            CACHE[task_name] = now

        if data:
            print ("")
        await asyncio.sleep(1)


asyncio.get_event_loop().run_until_complete(main())


status_collector/__init__.py
import traceback
import sys
import json
from decimal import Decimal
import decimal
import asyncio
import aiohttp
import aioredis
from aiohttp import ClientTimeout
from easyqueue import AsyncQueue, AsyncQueueConsumerDelegate
from status_collector import conf
import time
from aiologger.loggers.json import JsonLogger

timeout_config = ClientTimeout(
    connect=conf.STATS_CONNECT_TIMEOUT, total=conf.STATS_REQUEST_TIMEOUT
)


async def main():
    conf.cache = await aioredis.create_redis_pool(
        conf.REDIS_URL, minsize=conf.REDIS_POOL_MIN, maxsize=conf.REDIS_POOL_MAX
    )
    logger = JsonLogger.with_default_handlers(level=10, flatten=True)
    queue = AsyncQueue(
        conf.STATS_COLLECTOR_RABBITMQ_HOST,
        conf.STATS_COLLECTOR_RABBITMQ_USER,
        conf.STATS_COLLECTOR_RABBITMQ_PWD,
        delegate=Test(),
        virtual_host=conf.STATS_COLLECTOR_RABBITMQ_VHOST,
    )
    await fetch_app_stats(queue, logger)


async def get_slave_ip_list(master_ip, logger):
    try:
        session = aiohttp.ClientSession()
        resp = await session.get(
            f"http://{master_ip}:5050/slaves", timeout=timeout_config
        )
        data = await resp.json()
        await session.close()
        return list(
            map(lambda slave: f'{slave["hostname"]}:{slave["port"]}', data["slaves"])
        )
    except Exception as e:
        await session.close()
        await logger.exception(
            {
                "action": "get_slave_ip_list",
                "exception": {
                    "message": str(e),
                    "traceback": traceback.format_exc(),
                    "type": sys.exc_info()[0].__name__,
                },
                "master_ip": master_ip,
            }
        )
        raise e


def extract_app_name(task):
    app_name_with_namespace = task["executor_id"].rsplit(".", 1)[0]
    return "/" + app_name_with_namespace.replace("_", "/")


def extract_task_name(task):
    return task["executor_id"]


def round_up(n: Decimal, prec: int = 5) -> float:
    return float(n.quantize(Decimal("." + "0" * prec), rounding=decimal.ROUND_UP))


async def build_statistic_for_response(slave_ip, task_now):
    appname = extract_app_name(task_now)
    taskname = extract_task_name(task_now)

    task_before = await conf.cache.get(taskname)
    if not task_before:
        await conf.cache.set(taskname, json.dumps(task_now))
        return None

    task_before = json.loads(task_before)

    now_stats = task_now["statistics"]
    before_stats = task_before["statistics"]

    await conf.cache.set(taskname, json.dumps(task_now))

    period_secs = Decimal(now_stats["timestamp"] - before_stats["timestamp"])
    cpu_limit_raw = Decimal(str(now_stats["cpus_limit"]))
    cpu_limit = cpu_limit_raw - Decimal("0.1")

    cpu_thr_secs = Decimal(now_stats.get("cpus_throttled_time_secs", 0)) - Decimal(
        before_stats.get("cpus_throttled_time_secs", 0)
    )
    cpu_usr_secs = Decimal(now_stats["cpus_user_time_secs"]) - Decimal(
        before_stats["cpus_user_time_secs"]
    )
    cpu_sys_secs = Decimal(now_stats["cpus_system_time_secs"]) - Decimal(
        before_stats["cpus_system_time_secs"]
    )

    cpu_usr_host = cpu_usr_secs / period_secs
    cpu_sys_host = cpu_sys_secs / period_secs

    cpu_usr_host_pct = cpu_usr_host * 100
    cpu_sys_host_pct = cpu_sys_host * 100

    mem_bytes = now_stats["mem_rss_bytes"]
    mem_pct = (
        Decimal(now_stats["mem_rss_bytes"])
        / Decimal(now_stats["mem_limit_bytes"])
        * 100
    )

    cpu_usr_pct = cpu_usr_host_pct / cpu_limit
    cpu_sys_pct = cpu_sys_host_pct / cpu_limit
    cpu_thr_pct = cpu_thr_secs / period_secs * 100
    cpu_pct = (Decimal(cpu_usr_host) + Decimal(cpu_sys_host)) / Decimal(cpu_limit) * 100
    cpu_host_pct = cpu_usr_host_pct + cpu_sys_host_pct

    data = {
        "timestamp": now_stats["timestamp"],
        "cpu_limit": round_up(cpu_limit, prec=1),
        "cpu_limit_raw": round_up(cpu_limit_raw, prec=1),
        "host": slave_ip,
        "taskname": taskname,
        "appname": appname,
        "period_secs": round_up(period_secs),
        "cpu_usr_secs": round_up(cpu_usr_secs),
        "cpu_sys_secs": round_up(cpu_sys_secs),
        "cpu_thr_secs": round_up(cpu_thr_secs),
        "mem_bytes": mem_bytes,
        "mem_pct": round_up(mem_pct),
        "cpu_sys_host_pct": round_up(cpu_sys_host_pct),
        "cpu_usr_host_pct": round_up(cpu_usr_host_pct),
        "cpu_usr_pct": round_up(cpu_usr_pct),
        "cpu_sys_pct": round_up(cpu_sys_pct),
        "cpu_thr_pct": round_up(cpu_thr_pct),
        "cpu_pct": round_up(cpu_pct),
        "cpu_host_pct": round_up(cpu_host_pct),
    }

    # Para podermos diferencias tasks que não possuem os dados
    # necessários para calcular o throttling
    if "cpus_throttled_time_secs" not in now_stats:
        del data["cpu_thr_secs"]
        del data["cpu_thr_pct"]

    if conf.STATS_COLLECTOR_INCLUDE_RAW_METRICS:
        data["stats"] = {
            "before": task_before["statistics"],
            "now": task_now["statistics"],
        }

    return data


async def get_slave_statistics(slave_ip, logger):
    try:
        async with aiohttp.ClientSession() as session:
            async with session.get(
                f"http://{slave_ip}/monitor/statistics.json", timeout=timeout_config
            ) as resp:
                tasks = await resp.json()
                async_tasks = [
                    build_statistic_for_response(slave_ip, task) for task in tasks
                ]
    except Exception as e:
        await logger.exception(
            {
                "action": "get_slave_statistics",
                "exception": {
                    "message": str(e),
                    "traceback": traceback.format_exc(),
                    "type": sys.exc_info()[0].__name__,
                },
                "slave_ip": slave_ip,
            }
        )
        raise e

    results = await asyncio.gather(*async_tasks, return_exceptions=True)
    valid_results = []
    for r in results:
        if r and not isinstance(r, Exception):
            valid_results.append(r)
        if isinstance(r, Exception):
            await logger.exception(
                {
                    "action": "build_slave_stask_statistics",
                    "exception": {
                        "message": str(r),
                        "traceback": traceback.format_tb(r.__traceback__),
                        "type": r.__class__.__name__,
                    },
                    "slave_ip": slave_ip,
                }
            )

    return valid_results


async def send_slave_statistics_to_queue(slave_statistics, queue, logger):
    await queue.connect()
    for task in slave_statistics:
        await queue.put(body=task, routing_key=conf.STATS_COLLECTOR_RABBITMQ_RK)


async def async_tasks(ip, logger, queue):
    try:
        start = time.time()
        statistics = await get_slave_statistics(ip, logger)
        end = time.time()
        elapsed = end - start
        if statistics:
            await logger.info(
                {
                    "action": "fetch_slave_statistics",
                    "slaveIp": ip,
                    "totalTasks": len(statistics),
                    "processTime": elapsed,
                }
            )

            await send_slave_statistics_to_queue(statistics, queue, logger)
    except Exception as e:
        await logger.exception(e)


async def fetch_app_stats(queue, logger):
    start = time.time()
    ip_list = await get_slave_ip_list(conf.STATS_COLLECTOR_MESOS_MASTER_IP, logger)
    await logger.debug({"totalSlaves": len(ip_list), "slaveList": ip_list})
    try:
        tasks = [async_tasks(ip, logger, queue) for ip in ip_list]
        return_values = await asyncio.gather(*tasks, return_exceptions=True)
        end = time.time()
        elapsed = end - start
        await logger.info(
            {
                "action": "fetch_all_slaves_statistics",
                "totalSlaves": len(ip_list),
                "processTime": elapsed,
            }
        )
    except Exception as e:
        await logger.exception(e)


class Test(AsyncQueueConsumerDelegate):
    @property
    def queue_name(self) -> str:
        return "nothing"

    async def on_queue_message(self, content, delivery_tag, queue):
        pass
